music/src/ddsp/ddsp.ts

```diff
@@ -0,0 +1,387 @@
+      const ldMean = ldGreaterVal > 0 ? ldGreater.mean() : ldShifted.mean();
+
+      const ldDiffMean = tf.sub(SELECTED_LD_MEAN, ldMean);
+      const ldDiffMeanAdded = ldShifted.add(ldDiffMean);
+
+      const loudnessAdjustment = 0;
+      const ldAdjusted = ldDiffMeanAdded.add(loudnessAdjustment);
+
+      const ldClipped = ldAdjusted.clipByValue(LOWEST_LD, SELECTED_LD_AVG_MAX);
+
+      // rescale
+      const oldMin = ldClipped.min();
+      const subOldMin = ldMean.sub(oldMin);
+      return ldClipped
+        .sub(oldMin)
+        .div(subOldMin)
+        .mul(SELECTED_LD_MEAN - LOWEST_LD)
+        .add(LOWEST_LD);
+    });
+
+    const reshapedConf = tf.reshape(af.confidences, [-1, 1, 1]);
+    const smoothConf = tf.pool(
+      reshapedConf as Tensor3D,
+      [CONF_SMOOTH_SIZE, 1],
+      'avg',
+      'same'
+    );
+    const smoothConfReshaped = smoothConf.reshape([-1]);
+    const confMask = tf.lessEqual(smoothConfReshaped, CONF_THRESHOLD);
+
+    maskedPower = tf.tidy(() => {
+      const confReduction = confMask.mul(LD_CONF_REDUCTION);
+      const loudnessReduced = ldFinal.add(confReduction);
+      const finalMaskedPower = tf.maximum(loudnessReduced, LOWEST_LD);
+      confReduction.dispose();
+      loudnessReduced.dispose();
+      return finalMaskedPower;
+    });
+
+    loudness_db = (await maskedPower.array()) as number[];
+
+    ldShiftedGreater.dispose();
+    ldShifted.dispose();
+    ldFinal.dispose();
+    ldGreater.dispose();
+    smoothConf.dispose();
+    smoothConfReshaped.dispose();
+    reshapedConf.dispose();
+    confMask.dispose();
+  }
+  // tslint:disable-next-line: variable-name
+  let f0_hz: number[] = [];
+  if (af.f0_hz.length > 0) {
+    // shift the pitch register.
+    const _p = await hzToMidi(af.f0_hz);
+
+    const p = tf.tidy(() => {
+      for (let i = 0, n = _p.length; i < n; ++i) {
+        if (_p[i] === -Infinity) {
+          _p[i] = 0;
+        }
+      }
+
+      return _p;
+    });
+
+    const confMask = tf.lessEqual(af.confidences, CONF_THRESHOLD);
+    const maskedPowerThreshold = maskedPower.greater(SELECTED_LD_THRESH);
+    const mask = tf.logicalOr(confMask, maskedPowerThreshold);
+    const pMask = await tf.booleanMaskAsync(p, mask);
+    const pMean = pMask.mean();
+    const pDiff = await tf.sub(SELECTED_P_MEAN, pMean);
+    const pDiffVal = await pDiff.array();
+
+    const finalf0 = tf.tidy(() => {
+      let pDiffOctave = (pDiffVal as number) / 12;
+      pDiffOctave = Math.round(pDiffOctave);
+      const shiftedF0WithOctave = shiftF0(af.f0_hz, pDiffOctave);
+      return shiftedF0WithOctave;
+    });
+
+    f0_hz = (await finalf0.array()) as number[];
+
+    maskedPower.dispose();
+    pMean.dispose();
+    pMask.dispose();
+    pDiff.dispose();
+    finalf0.dispose();
+    maskedPowerThreshold.dispose();
+    mask.dispose();
+    confMask.dispose();
+  }
+
+  return { f0_hz, loudness_db };
+}
+
+export function getModelValues(checkpointUrl: string): ModelValues {
```

Comment:
This is part of that leaky abstraction, and should probably take place in the model class constructor.